### PR TITLE
Add testing and tooling for DEX integrations

### PR DESCRIPTION
**Pull request by shresthagrawal**
Fixes BACK-428

### CODE REVIEW

**Review by ColonelJ — CHANGES_REQUESTED** [review]

File: scripts/fetch-event-blocknumber.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+import { Contract } from '@ethersproject/contracts';
+import { JsonRpcProvider } from '@ethersproject/providers';
+import VaultABI from '../src/abi/balancer-v2/vault.json';
+import { ProviderURL, Network } from '../src/constants';
+import { Address } from '../src/types';
+
+// This is a helper script to fetch blocknumbers where a certain
+// event was released by a certain contract
+async function getBlockNumbersForEvents(
```
[the remark this answers is not in the document]
Comment (shresthagrawal) [reply scripts/fetch-event-blocknumber.ts:12]: This is just a script that devs can modify locally to fetch the blockNumbers. Not sure if its required to commit it as a separate script to the repo

File: src/dex/uniswap-v2/optimizer.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import { UniswapV2Config } from './config';
+import { MDEXConfig } from './mdex';
+import { BiSwapConfig } from './biswap';
+import _ from 'lodash';
+
+const AllUniswapForks = [
+  ...Object.keys(UniswapV2Config),
+  ...Object.keys(MDEXConfig),
+  ...Object.keys(BiSwapConfig),
+];
```
[the remark this answers is not in the document]
Comment (shresthagrawal) [reply src/dex/uniswap-v2/optimizer.ts:12]: I will simply remove the BakerySwap from the optimiser for now

File: .gitignore [hunk 3]
```diff
@@ -2,4 +2,6 @@ node_modules
 .DS_Store
 build/
 yarn-error.log
-.env
\ No newline at end of file
+.env
+tests/logs.json
+tests/states.json
```
[the remark this answers is not in the document]
Comment (shresthagrawal) [reply .gitignore:7]: fixed

File: src/constants.ts [hunk 4]
```diff
@@ -39,3 +39,25 @@ export const MULTI_V2: { [network: number]: Address } = {
   [Network.AVALANCHE]: '0xd7Fc8aD069f95B6e2835f4DEff03eF84241cF0E1',
   [Network.FANTOM]: '0xdC6E2b14260F972ad4e5a31c68294Fba7E720701',
 };
+
+export const ProviderURL: { [network: number]: string } = {
+  [Network.MAINNET]: process.env.HTTP_PROVIDER_1 || '',
```
[the remark this answers is not in the document]
Comment (shresthagrawal) [reply src/constants.ts]: fixed

File: src/dex/uniswap-v2/uniswap-v2-e2e.test.ts [hunk 5]
```diff
@@ -0,0 +1,409 @@
+import dotenv from 'dotenv';
```
[the remark this answers is not in the document]
Comment (shresthagrawal) [reply src/dex/uniswap-v2/uniswap-v2-e2e.test.ts:1]: added

File: tests/tenderly-simulation.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      );
+      this.forkId = res.data.simulation_fork.id;
+      this.lastTx = res.data.root_transaction.id;
+    } catch (e) {
+      console.error(`TenderlySimulation_setup: ${e.message}`, e.trace);
+      throw e;
+    }
+  }
+
+  async simulate(params: any) {
```
[the remark this answers is not in the document]
Comment (shresthagrawal) [reply tests/tenderly-simulation.ts]: fixed

File: tests/utils-e2e.ts [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+
+  const swapTx = await ts.simulate(swapParams);
+  console.log(`${srcToken.address}_${destToken.address}_${dexKey!}`);
+  // Only log gas estimate if testing against API
+  // if (testingEndpoint)
+  //   console.log(
+  //     `Gas Estimate API: ${priceRoute.gasCost}, Simulated: ${
+  //       swapTx!.gasUsed
+  //     }, Difference: ${priceRoute.gasCost - swapTx!.gasUsed}`,
+  //   );
```
[the remark this answers is not in the document]
Comment (shresthagrawal) [reply tests/utils-e2e.ts]: done

**Review by ColonelJ — APPROVED** [review]
Probably best to merge to the other PR, and deal with any further improvements there.

**Pull request merged**